src/Microsoft.AspNetCore.SignalR.Client/RpcConnection.cs

```diff
@@ -0,0 +1,245 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.IO.Pipelines;
+using System.Linq;
+using System.Net.Http;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Sockets.Client;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.AspNetCore.SignalR.Client
+{
+    public class HubConnection : IDisposable
+    {
+        private readonly Task _reader;
+        private readonly Stream _stream;
+        private readonly ILogger _logger;
+        private readonly Connection _connection;
+        private readonly IInvocationAdapter _adapter;
+        private readonly HubBinder _binder;
+
+        private readonly CancellationTokenSource _readerCts = new CancellationTokenSource();
+        private readonly ConcurrentDictionary<string, InvocationRequest> _pendingCalls = new ConcurrentDictionary<string, InvocationRequest>();
+        private readonly ConcurrentDictionary<string, InvocationHandler> _handlers = new ConcurrentDictionary<string, InvocationHandler>();
+
+        private int _nextId = 0;
+
+        private HubConnection(Connection connection, IInvocationAdapter adapter, ILogger logger)
+        {
+            _binder = new HubBinder(this);
+            _connection = connection;
+            _stream = connection.GetStream();
+            _adapter = adapter;
+            _logger = logger;
+
+            _reader = ReceiveMessages(_readerCts.Token);
+        }
+
+        // TODO: Client return values/tasks?
+        // TODO: Overloads that use type parameters (like On<T1>, On<T1, T2>, etc.)
+        public void On(string methodName, Type[] parameterTypes, Action<object[]> handler)
```

Comment:
Yea, we're going to talk about this pattern. I think we'll fundamentally move away from this.